content/posts/side-effects.md

```diff
@@ -0,0 +1,113 @@
+
+Everything else is considered a _side effect_ e.g. sending a message, POSTing to a HTTP endpoint, creating a blob for the PDF invoice document etc. So given that there can be no atomic transaction between the targets of these side effects and the transactional store, how can we ensure correctness? We have seen in [the previous post](https://exactly-once.github.io/posts/token-based-deduplication/) that it can be done for outgoing message so let's try to apply the same technique in a more general way.
+
+Let's think about requirements for correctly applying the side effects. Fist, **we want all side effects to be visible before the incoming message is consumed**. That condition includes that we of course don't want them to be omitted. Next, **we don't want any side effects to be visible if the incoming message have not been processed** e.g. because of processing logic throwing an exception. 
+
+Third condition is less obvious. Because we need to always take into account possibility of attempting to process the same logical message concurrently by multiple threads, we need to ensure that the **side effects that are published all come from the same processing attempt**. For example if the message handler creates two blobs, we don't want to allow situation where each blob is created by different processing attempt as this could lead to inconsistent state if these handlers were not purely deterministic. 
+
+Fourth and the last condition is about efficiency of storage usage. We don't want to crate any garbage so **by the time the message is processed successfully (consumed), all side effects that resulted from non-successful attempts should have been cleaned up**.
+
+### Sketch of an algorithm
+
+Looking at the above conditions we can try to sketch an algorithm. In [the previous approach](https://exactly-once.github.io/posts/token-based-deduplication/) we had our side effects (outgoing messages) serialized and stored in the transactional store alongside the business data. This is not a great approach if we want to support generalized side effects as they can be of significant size. A PFD document we want to store or send via e-mail can easily weigh few megabytes. There is no way we can include things like this as part of a document stored in CosmosDB.
+
+That means we need to actually create side effects as we execute the handler and not store them serialized. But how about our second rule of not having them visible before the message is processed? Here's the trick. **Visible is not the same as created**. We can create a blob but before we tell about it, there is no side effect as far as other endpoints of the system are concerned.
```

Comment:
```suggestion
That means we need to create side effects as the handler is executed rather than serializing and storing them is a persistent store. But how about our second rule of not having them visible before the message is processed? Here's the trick. **Visible is not the same as created**. We can create a blob but before we tell about it, there is no side effect as far as other endpoints of the system are concerned.
```